code/network/lib/Loot/Network/ZMQ/Client.hs

```diff
@@ -0,0 +1,270 @@
+
+-- | Client environment state. Is to be used by the one thread only
+-- (main client worker).
+data ZTNetCliEnv = ZTNetCliEnv
+    {
+      ztCliBack         :: Z.Socket Z.Router
+      -- ^ Backend which receives data from the network and routes it
+      -- to client workers.
+    , ztCliSub          :: Z.Socket Z.Sub
+      -- ^ Subscriber socket which is listening to other nodes'
+      -- updates. It also sends data to clients .
+
+    , ztPeers           :: TVar (Set ZTNodeId)
+      -- ^ List of peers we are connected to. Is to be updated by the
+      -- main thread only (broker worker).
+
+    , ztClients         :: TVar (Map ClientId ZTClientEnv)
+      -- ^ Channels binding broker to clients, map from client name to.
+    , ztSubscriptions   :: TVar (Map Subscription (Set ClientId))
+      -- ^ Map from subscription key to clents' identifiers.
+    , ztMsgTypes        :: TVar (Map MsgType ClientId)
+      -- ^ Map from msg type key to clents' identifiers.
+
+    , ztCliRequestQueue :: CliRequestQueue
+      -- ^ Queue to read (internal, administrative) client requests
+      -- from, like updating peers or resetting connection.
+    }
+
+instance HasLens ZTNetCliEnv r ZTNetCliEnv =>
+         HasLens CliRequestQueue r CliRequestQueue where
+    lensOf =
+        (lensOf @ZTNetCliEnv) .
+        (lens ztCliRequestQueue (\ztce rq2 -> ztce {ztCliRequestQueue = rq2}))
+
+createNetCliEnv :: MonadIO m => ZTGlobalEnv -> Set ZTNodeId -> m ZTNetCliEnv
+createNetCliEnv (ZTGlobalEnv ctx) peers = liftIO $ do
+    -- I guess it's alright to connect ROUTER instead of binding it.
+    -- https://stackoverflow.com/questions/16109139/zmq-when-to-use-zmq-bind-or-zmq-connect
+    ztCliBack <- Z.socket ctx Z.Router
+    forM_ peers $ Z.connect ztCliBack . ztNodeIdRouter
+
+    ztCliSub <- Z.socket ctx Z.Sub
+
+    ztClients <- newTVarIO mempty
+    ztPeers <- newTVarIO peers
+    ztSubscriptions <- newTVarIO mempty
+    ztMsgTypes <- newTVarIO mempty
+    ztCliRequestQueue <- CliRequestQueue <$> TQ.newTQueueIO
+
+    pure ZTNetCliEnv {..}
+
+----------------------------------------------------------------------------
+-- Methods
+----------------------------------------------------------------------------
+
+data CliBrokerStmRes
+    = CBClient ClientId (Maybe ZTNodeId) (MsgType, Content)
+    | CBBack
+    | CBSub
+    | CBRequest InternalRequest
+
+runBroker :: (MonadReader r m, HasLens' r ZTNetCliEnv, MonadIO m, MonadMask m) => m ()
+runBroker = do
+    ZTNetCliEnv{..} <- view $ lensOf @ZTNetCliEnv
+    -- This function may do something creative (LRU! Lowest ping!),
+    -- but instead (for now) it'll just choose a random peer.
+    let choosePeer = do
+            -- Yes, modulo bias. It shouldn't really matter.
+            i <- abs <$> randomIO
+            atomically $ do
+                l <- Set.toList <$> readTVar ztPeers
+                pure $ l L.!! (i `mod` length l)
```

Comment:
I don't think it matters here, but doing `modulus N` on some random number when `N` is not `2^M` actually creates a probability spike.

Reply:
I don't quite get the criticism -- haven't you noticed the comment 5 lines higher?

Also, what? I suppose you meant `i` should have been in `[0,k]`, where `k = 0 (mod (length l))`, but (again), for small lists it's totally negligible.

Reply:
Oh, okay then. I didn't see that comment.